src/mfa/webauthn/controller/attestation.ts

```diff
@@ -0,0 +1,95 @@
+import Controller from '@curveball/controller';
+import { Context } from '@curveball/core';
+import { generateAttestationOptions, verifyAttestationResponse } from '@simplewebauthn/server';
+
+import * as webAuthnService from '../service';
+import { getSetting } from '../../../server-settings';
+import { User } from '../../../user/types';
+
+
+class WebAuthnAttestationController extends Controller {
+
+  async get(ctx: Context) {
+    const user: User = ctx.state.session.registerUser;
+
+    if (!getSetting('webauthn.relyingPartyId')) {
+      console.error(`Error: The webauthn.relyingPartyId server-setting is not set. This should match the domain the page is served from (ex: login.example.com).
+See the Relying Party Identifier section of the WebAuthn W3C Recommendation here:
+https://www.w3.org/TR/webauthn/#rp-id`);
+    }
+
+    const attestationOptions = generateAttestationOptions({
+      serviceName: getSetting('webauthn.serviceName'),
+      rpID: getSetting('webauthn.relyingPartyId'),
+      userID: user.id.toString(),
+      userName: user.nickname,
+      timeout: 60000,
+      attestationType: 'direct',
+      excludedCredentialIDs: (await webAuthnService.findDevicesByUser(user)).map(device => device.credentialID),
+      /**
+       * The optional authenticatorSelection property allows for specifying more constraints around
+       * the types of authenticators that users to can use for attestation
+       */
+      authenticatorSelection: {
+        authenticatorAttachment: 'cross-platform',
+        userVerification: 'preferred',
+        requireResidentKey: false,
+      },
+    });
+
+    ctx.state.session.webAuthnChallengeRegister = attestationOptions.challenge;
+    ctx.response.body = attestationOptions;
+  }
+
+  async post(ctx: Context) {
+    const user: User = ctx.state.session.registerUser;
+    const body = ctx.request.body;
+
+    const expectedChallenge = ctx.state.session.webAuthnChallengeRegister;
+    ctx.state.session.webAuthnChallengeRegister = null;
+
+    if (!getSetting('webauthn.expectedOrigin')) {
+      console.error(`Error: The webauthn.expectedOrigin server-setting is not set. This should match the origin the browser is making the request from (ex: https://login.example.com)
+See the Relying Party Identifier section of the WebAuthn W3C Recommendation here:
+https://www.w3.org/TR/webauthn/#rp-id`);
```

Comment:
What's expected to happen if this setting is not set? Can we still continue?

Reply:
If this is not set or the setting is incorrect, the server will return an error to the client. It still "works", but clients will receive errors until the correct origin is set. The server is expected to validate that the origin of the client that is sending the request matches the origin that it expects. This is to prevent clients from registering devices from origins the client doesn't expect.

The client error looks like this when registering a device:
![Screenshot from 2020-08-13 21-26-51](https://user-images.githubusercontent.com/11312558/90202450-bf7ce180-ddab-11ea-8b32-5578f2dbd6eb.png)
`http://bogus-localhost:8531` is the server setting.

This setting is also used for logging in (assertion). The client gets a similar error:
![Screenshot from 2020-08-13 21-20-16](https://user-images.githubusercontent.com/11312558/90202187-01595800-ddab-11ea-8796-d1f0e45a26ba.png)

Reply:
It's number 2 in Step 6 from `Registration` described here:
https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API

Reply:
Ah I guess I just got confused because there was no `throw` or `return`..

Awesome!